lib/checkother.cpp

```diff
@@ -1933,6 +1933,34 @@ void CheckOther::checkDuplicateExpression()
             continue;
 
         for (const Token *tok = scope->classStart; tok && tok != scope->classEnd; tok = tok->next()) {
+            if(tok->str() == "=" && Token::Match(tok->astOperand1(), "%var%")) {
+                const Token * endStatement = Token::findsimplematch(tok, ";");
+                if(Token::Match(endStatement, "; %type% %var% ;")) {
+                    endStatement = endStatement->tokAt(4);
+                }
+                if(Token::Match(endStatement, "%var% %assign%")) {
+                    const Token * nextAssign = endStatement->tokAt(1);
+                    const Token * var1 = tok->astOperand1();
+                    const Token * var2 = nextAssign->astOperand1();
+                    if(Token::Match(var1->previous(), ";|{|} %var%") &&
+                        Token::Match(var2->previous(), ";|{|} %var%") &&
+                        var2->valueType() && var1->valueType() && 
+                        var2->valueType()->originalTypeName == var1->valueType()->originalTypeName &&
+                        var2->valueType()->pointer == var1->valueType()->pointer &&
+                        var2->valueType()->constness == var1->valueType()->constness &&
+                        var2->varId() != var1->varId() && 
+                        !tok->next()->isLiteral() &&
```

Comment:
you check that the first token in rhs is not a literal? so you don't want to warn about:

    x = 1 + y;

maybe I am missing something?

I would recommend that you use the AST entirely. don't look at the first token in rhs.

If you don't want that RHS is a number or name .. I would write something like:

    Token::Match(tok->astOperand2(), "%name%|%num%")

You don't need to have both `tok->next()->isLiteral()` and `nextAssign->next()->isLiteral()` if you also use `isSameExpression` later. If they are "same" then either both are literals or not literals.

Reply:
> you check that the first token in rhs is not a literal?

Its to avoid warning about expressions like:

```cpp
int x = 0;
int y = 0;
```

But I see that I need to make it better as it will skip `x = 1 + y;` as you mentioned.

> If I wanted that RHS was some arithmetic expression or function call I would try

That does sound better. Let me try that.